dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java

```diff
@@ -0,0 +1,126 @@
+
+/**
+ * Tests that {@link GwtAstBuilder} correctly builds the AST for
+ * features introduced in Java 17.
+ */
+public class Java17AstTest extends FullCompileTestBase {
+
+  @Override
+  public void setUp() throws Exception {
+    super.setUp();
+    addAll(JavaResourceBase.createMockJavaResource("test.TextBlock",
+        "package test;",
+        "public interface TextBlock {",
+        "String text =\"\"\"",
+        "line 1",
+        "line 2",
+        "line 3",
+        "\"\"\";",
+        "}"
+    ));
+
+    addAll(JavaResourceBase.createMockJavaResource("test.Shape",
+        "package test;",
+        "public sealed class Shape permits Square, Circle {",
+        "}"
+    ));
+
+    addAll(JavaResourceBase.createMockJavaResource("test.Square",
+        "package test;",
+        "public final class Square extends Shape {",
+        "}"
+    ));
+
+    addAll(JavaResourceBase.createMockJavaResource("test.Circle",
+        "package test;",
+        "public final class Circle extends Shape {",
+        "}"
+    ));
+
+    addAll(JavaResourceBase.createMockJavaResource("test.Months",
+        "package test;",
+        "public enum Months {",
+        "JANUARY, FEBRUARY, MARCH, APRIL, MAY, JUNE, JULY, AUGUST, SEPTEMBER, OCTOBER, NOVEMBER, DECEMBER;",
+        "}"
+    ));
+  }
+
+  public void testTextBlocks() throws Exception {
+    JProgram program = compileSnippet("void", "new TextBlock(){};");
+    JInterfaceType textBlock = (JInterfaceType) findType(program, "TextBlock");
+    JStringLiteral initializer = (JStringLiteral) textBlock.getFields().get(0).getInitializer();
+    String multiLineString = initializer.getValue();
+    List<String> lines = Arrays.asList(multiLineString.split("\n"));
+    assertEquals(3, lines.size());
+    assertEquals("line 1", lines.get(0));
+    assertEquals("line 2", lines.get(1));
+    assertEquals("line 3", lines.get(2));
+  }
+
+  public void testSealedClassesPermitted() throws Exception {
+    compileSnippet("void", "Shape square = new Square();");
+    compileSnippet("void", "Shape circle = new Circle();");
+  }
+
+  public void testSealedClassesNotPermitted() {
+    try {
+      addSnippetClassDecl("public final class Rectangle extends Shape {" +
+          "}");
+      compileSnippet("void", "Shape rectangle = new Rectangle();");
+      fail("Compile should have failed but succeeded.");
+    } catch (Exception e) {
+      if (!(e.getCause() instanceof UnableToCompleteException)
+          && !(e instanceof UnableToCompleteException)) {
+        e.printStackTrace();
+        fail();
+      }
+    }
+  }
+
+  public void testSwitchExpressionsNotSupported() {
+    try {
+      compileSnippet("void", "var month = Months.JUNE;" +
+          "var result = switch(month) {\n" +
+          "    case JANUARY, JUNE, JULY -> 3;\n" +
+          "    case FEBRUARY, SEPTEMBER, OCTOBER, NOVEMBER, DECEMBER -> 1;\n" +
+          "    case MARCH, MAY, APRIL, AUGUST -> 2;\n" +
+          "    default -> 0;" +
```

Comment:
aside: what's the pattern here? J means 3, M and A means 2, everything else is 1?

Reply:
It is just a random pattern for testing, didnt need to mean anything.